Marcus Diehl
PGH Heathly Ride

Starting off with the importation of Pandas

In [1]:
import pandas as pd

Broken into parts to confirm I'm getting everything
here we scan for CSV files from a parent folder and all roots

In [12]:
import os

parent_dir = '.'
subject_dirs = [os.path.join(parent_dir, dir) for dir in os.listdir(parent_dir) if os.path.isdir(os.path.join(parent_dir, dir))]

filelist = []
for dir in subject_dirs:
    csv_files = [os.path.join(dir, csv) for csv in os.listdir(dir) if os.path.isfile(os.path.join(dir, csv)) and csv.endswith('.csv')]
    for file in csv_files:
        filelist.append(file)
#rentals_df = pd.read_csv('rentals.csv', encoding='latin-1')

In [13]:
filelist

['.\\2015-Q2\\HealthyRide Rentals 2015 Q2.csv',
 '.\\2015-Q2\\HealthyRideStations2015.csv',
 '.\\2015-Q3\\HealthyRide Rentals 2015 Q3.csv',
 '.\\2015-Q3\\HealthyRideStations2015.csv',
 '.\\2015-Q4\\HealthyRide Rentals 2015 Q4.csv',
 '.\\2015-Q4\\HealthyRideStations2015.csv',
 '.\\2016-Q1\\HealthyRide Rentals 2016 Q1.csv',
 '.\\2016-Q1\\HealthyRideStations2016.csv',
 '.\\2016-Q2\\HealthyRide Rentals 2016 Q2.csv',
 '.\\2016-Q2\\HealthyRideStations2016.csv',
 '.\\2016-Q3\\HealthyRide Rentals 2016 Q3.csv',
 '.\\2016-Q3\\HealthyRideStations2016.csv',
 '.\\2016-Q4\\HealthyRide Rentals 2016 Q4.csv',
 '.\\2016-Q4\\HealthyRideStations2016.csv',
 '.\\2017-Q1\\Healthy Ride Rentals 2017-Q1.csv',
 '.\\2017-Q1\\HealthyRideStations2017.csv',
 '.\\2017-Q2\\Healthy Ride Rentals 2017-Q2.csv',
 '.\\2017-Q2\\HealthyRideStations2017.csv']

It worked, but we've grabbed all the 'Stations' files as well, and don't need those just yet
But interesting that the file names for the stations don't include the quarter, just the year.

So now exclude the stations and read the contents into a single Panda DataFrame

In [7]:
rentals_df = pd.DataFrame()
list_ = []
for file_ in filelist:
    if 'Stations' not in file_:
        df = pd.read_csv(file_,index_col=None, header=0)
        list_.append(df)
rentals_df = pd.concat(list_)

In [15]:
rentals_df.head()

Trip id       Starttime         Stoptime  Bikeid  Tripduration  \
0  10676972  5/31/2015 6:54  5/31/2015 10:17   70166         12190   
1  10677086  5/31/2015 6:57   5/31/2015 9:24   70368          8842   
2  10677131  5/31/2015 6:58  5/31/2015 10:09   70354         11448   
3  10677525  5/31/2015 7:08   5/31/2015 7:15   70232           452   
4  10677886  5/31/2015 7:15   5/31/2015 9:27   70468          7866   

   From station id           From station name  To station id  \
0             1018         37th St & Butler St           1001   
1             1018         37th St & Butler St           1018   
2             1019         42nd St & Butler St           1019   
3             1001  Forbes Ave & Market Square           1001   
4             1018         37th St & Butler St           1018   

              To station name Usertype  
0  Forbes Ave & Market Square      NaN  
1         37th St & Butler St      NaN  
2         42nd St & Butler St      NaN  
3  Forbes Ave & Market Square      NaN  
4         37th St & Butler St      NaN

well that worked, but my concern is did I repeat the header row 12 times as internal content
Time will tell when I start the Mean time analysis work
Something you'll notice that in 2015, they didn't have the CustomerType field populated

So what are we working with?


In [17]:
rentals_df.dtypes

Trip id               int64
Starttime            object
Stoptime             object
Bikeid                int64
Tripduration          int64
From station id       int64
From station name    object
To station id         int64
To station name      object
Usertype             object
dtype: object

based on on the Object types, I'm guessing the headers didn't get replicated or we would'nt have any Int64's

In [18]:
rentals_df.shape

(170531, 10)

170,531 rows and 10 columns


As agreed upon for the first qustion to answer, we wanted to know the Mean Time for ride duration grouped by UserType
At first I thought I had to conver the Timestamps from string to Date and then do a different in time caluclation to compare Start to Stop, but lookslike Trip duration is already included as a column
Tripduration is in Seconds

The hypotheis is, Subscribers are by definition more frequent users of the Bike share and thus have a purpose while Customers are enjoying the bikes and are more leasurely. So I expect Customers to hold the bikes for longer averages than Subscribers

So lets just branch out into a new Variable with  usertype and Tripduration columns

In [27]:
tripmeantime_df = rentals_df[['Usertype','Tripduration']]

now to find the mean duration by User Type.... 

Sort of found the answer here, though they wanted like a double average..
https://stackoverflow.com/questions/30328646/python-pandas-group-by-in-group-by-and-average


In [33]:
tripmeantime_df.groupby(['Usertype']).mean()

Tripduration
Usertype                
Customer     4921.921484
Subscriber   1230.911230